winch/codegen/src/codegen/mod.rs

```diff
@@ -348,6 +348,17 @@ where
     /// the given index, returning the typed register holding the
     /// source value.
     pub fn emit_set_local(&mut self, index: u32) -> TypedReg {
+        // Materialize any references to the same local index that are in the
+        // value stack by spilling.
+        let stack_contains_latent_local = self
+            .context
+            .stack
+            .iter()
+            .skip(1) // top-most element is popped below.
+            .any(|v| v.is_local_at_index(index));
```

Comment:
We talked offline.

We can use an `iter().rev().take_while(|v| /* v is not mem or a const */)` approach to avoid iterating over the entire stack. And the stack method I'm adding will be named in such a way that it indicates it does not check the top element. This is because we can safely avoid spilling if it's only the top element that references the local since we're going to pop the top element immediately after performing the check to see if the stack contains the local.